pkg/services/object/acl/basic.go

```diff
@@ -0,0 +1,398 @@
+	if err != nil {
+		return nil, err
+	}
+
+	reqInfo, err := b.findRequestInfo(request, cid, acl.OperationRange)
+	if err != nil {
+		return nil, err
+	}
+
+	if !basicACLCheck(reqInfo) {
+		return nil, ErrBasicAccessDenied
+	}
+
+	stream, err := b.next.GetRange(ctx, request)
+	return getRangeStreamBasicChecker{stream}, err
+}
+
+func (b BasicChecker) GetRangeHash(
+	ctx context.Context,
+	request *object.GetRangeHashRequest) (*object.GetRangeHashResponse, error) {
+
+	cid, err := getContainerIDFromRequest(request)
+	if err != nil {
+		return nil, err
+	}
+
+	reqInfo, err := b.findRequestInfo(request, cid, acl.OperationRangeHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if !basicACLCheck(reqInfo) {
+		return nil, ErrBasicAccessDenied
+	}
+
+	return b.next.GetRangeHash(ctx, request)
+}
+
+func (p putStreamBasicChecker) Send(request *object.PutRequest) error {
+	body := request.GetBody()
+	if body == nil {
+		return ErrMalformedRequest
+	}
+
+	part := body.GetObjectPart()
+	if _, ok := part.(*object.PutObjectPartInit); ok {
+		cid, err := getContainerIDFromRequest(request)
+		if err != nil {
+			return err
+		}
+
+		owner, err := getObjectOwnerFromMessage(request)
+		if err != nil {
+			return err
+		}
+
+		reqInfo, err := p.source.findRequestInfo(request, cid, acl.OperationPut)
+		if err != nil {
+			return err
+		}
+
+		if !basicACLCheck(reqInfo) || !stickyBitCheck(reqInfo, owner) {
+			return ErrBasicAccessDenied
+		}
+	}
+
+	return p.next.Send(request)
+}
+
+func (p putStreamBasicChecker) CloseAndRecv() (*object.PutResponse, error) {
+	return p.next.CloseAndRecv()
+}
+
+func (g getStreamBasicChecker) Recv() (*object.GetResponse, error) {
+	resp, err := g.next.Recv()
+	if err != nil {
+		return resp, err
+	}
+
+	body := resp.GetBody()
+	if body == nil {
+		return resp, err
+	}
+
+	part := body.GetObjectPart()
+	if _, ok := part.(*object.GetObjectPartInit); ok {
+		owner, err := getObjectOwnerFromMessage(resp)
+		if err != nil {
+			return nil, err
+		}
+
+		if !stickyBitCheck(g.info, owner) {
+			return nil, ErrBasicAccessDenied
+		}
+	}
+
+	return resp, err
+}
+
+func (b BasicChecker) findRequestInfo(
+	req RequestV2,
+	cid *refs.ContainerID,
+	op acl.Operation) (info requestInfo, err error) {
+
+	// container.Source interface implemented with SDK's definitions,
+	// so we have to convert id there.
+	containerID := containerSDK.NewIDFromV2(cid)
+
+	// fetch actual container
+	cnr, err := b.containers.Get(containerID)
+	if err != nil || cnr.GetOwnerID() == nil {
+		return info, ErrUnknownContainer
+	}
+
+	// find request role
+	role := b.sender.Classify(req, cid, cnr)
+	if role == acl.RoleUnknown {
+		return info, ErrUnknownRole
+	}
+
+	info.basicACL = cnr.GetBasicACL()
+	info.requestRole = role
+	info.operation = op
+	info.owner = cnr.GetOwnerID()
+
+	return info, nil
+}
+
+func getContainerIDFromRequest(req interface{}) (id *refs.ContainerID, err error) {
+	defer func() {
+		// if there is a NPE on get body and get address
```

Comment:
Can it really arise?

Reply:
Sure, one can send request with `nil` owner id. I don't mind to have panic recovery directly in network/transport/object/grpc though. In this case recovery there can be omitted. What do you think?

Reply:
Current code inside function body does not seem to panic cause of this since used getters are NPE-safe.

Reply:
They are not NPE-safe unfortunately

```
func (r *GetRequestBody) GetAddress() *refs.Address {
	if r != nil {
		return r.addr
	}

	return nil
}
```

Reply:
```
func TestNPE(t *testing.T) {
	var req *object2.GetRequest

	require.NotPanics(t, func() {
		req.GetBody().GetAddress().GetContainerID().GetValue()
	})
}

--- PASS: TestNPE (0.00s)
PASS
```

Reply:
Oh, I see, they all have that nil check inside. Okay, I'll remove this recovery but still wonder about recovery at network/transport/object/grpc just in case.

Reply:
Fixed in https://github.com/nspcc-dev/neofs-node/pull/32/commits/6812073152511f371bc572397518743cc790cd97